biogeochem/FatesLandUseChangeMod.F90

```diff
@@ -0,0 +1,307 @@
+       write(fates_log(),*) 'ruleset: ', ruleset
+       call endrun(msg=errMsg(sourcefile, __LINE__))
+
+    end select
+
+  end subroutine get_landusechange_rules
+
+  !----------------------------------------------------------------------------------------------------
+
+  subroutine get_luh_statedata(bc_in, state_vector)
+
+    type(bc_in_type) , intent(in) :: bc_in
+    real(r8), intent(out) :: state_vector(n_landuse_cats)  ! [m2/m2]
+
+    ! LOCALS
+    type(luh2_fates_lutype_map) :: lumap
+    real(r8) :: temp_vector(hlm_num_luh2_states)  ! [m2/m2]
+    real(r8) :: urban_fraction
+    integer  :: i_luh2_states
+    integer  :: ii
+    character(5) :: state_name
+    logical :: modified_flag
+
+    ! zero state vector and urban fraction
+    state_vector(:) = 0._r8
+    urban_fraction = 0._r8
+
+    ! Check to see if the incoming state vector is NaN.
+    temp_vector = bc_in%hlm_luh_states
+    call CheckLUHData(temp_vector,modified_flag)
+    if (.not. modified_flag) then
+       ! identify urban fraction so that it can be factored into the land use state output
+       urban_fraction = bc_in%hlm_luh_states(findloc(bc_in%hlm_luh_state_names,'urban',dim=1))
+    end if
+
+    ! loop over all states and add up the ones that correspond to a given fates land use type
+    do i_luh2_states = 1, hlm_num_luh2_states
+
+       ! Get the luh2 state name and determine fates aggregated land use
+       ! type index from the state to lutype map
+       state_name = bc_in%hlm_luh_state_names(i_luh2_states)
+       ii = lumap%GetIndex(state_name)
+
+       ! Avoid 'urban' states whose indices have been given unset values
+       if (ii .ne. fates_unset_int) then
+          state_vector(ii) = state_vector(ii) + &
+               temp_vector(i_luh2_states) / (1._r8 - urban_fraction)
+       end if
+    end do
+
+    ! check to ensure total area == 1, and correct if not
+    if ( abs(sum(state_vector(:)) - 1._r8) .gt. nearzero ) then
+       write(fates_log(),*) 'warning: sum(state_vector) = ', sum(state_vector(:))
+       state_vector = state_vector / sum(state_vector)
+    end if
+
+  end subroutine get_luh_statedata
+
+  !----------------------------------------------------------------------------------------------------
+
+  subroutine CheckLUHData(luh_vector,modified_flag)
+
+    use shr_infnan_mod   , only : isnan => shr_infnan_isnan
+
+    real(r8), intent(inout) :: luh_vector(:)  ! [m2/m2]
+    logical, intent(out)    :: modified_flag
+
+    ! Check to see if the incoming luh2 vector is NaN.
+    ! This suggests that there is a discepency where the HLM and LUH2 states
+    ! there is vegetated ground. E.g. LUH2 data is missing for glacier-margin regions such as Antarctica.
+    ! In this case, states should be Nan.  If so,
+    ! set the current state to be all primary forest, and all transitions to be zero.
+    ! If only a portion of the vector is NaN, there is something  amiss with
+    ! the data, so end the run.
+
```

Comment:
@ckoven I can't tell if there are situations here where we should be gracefully failing, instead of pushing on.  Can you confirm that the only possibility here is glacier intervention? Is this glacier retreat?

It seems like if we have a nan, something is wrong, no?  Are there any scenarios where a nan would be worth aborting? Like what if we have a precedent on previous steps with valid transitions?

Reply:
The issue is just that there are places where luh2 doesn't cover, e.g., all of Antarctica even though not all of Antarctica is ice-covered. So we don't want it to fail there, we just want to not use the land use data, and instead just assume that there is no land use at that point (and thus everything is primary). I think the subsequent logic handles the case if something has gone wrong somehow in the luh2 data for a given site and it is NaN somewhere but not everywhere.